### ISSUE TITLE
Redirect for reauth should not present login form

### ISSUE THREAD

**Issue by janus-reith** [reactioncommerce/example-storefront#661]
## Issue Description
Currently, when the storefront recieves a Status 401 from the backend over GraphQL, it will redirect to the `/signin` route. This will then redirect to the meteor backend to get a new authentication token. The assumption here seems to be, that the Meteor User still is logged in. If the User is logged out on the Backend, this will just present a login form. Skipping the auth redirect won't solve anything, as the page will just stay at its current point and display a Status 500 then.

[example-storefront/src/lib/apollo/initApollo.js](https://github.com/reactioncommerce/example-storefront/blob/ff29b3ec8bc72e9034056b90fd5a4940e3bbf348/src/lib/apollo/initApollo.js#L50)

The desired behaviour should be: If GraphQL gives a 401, get an actual indication if there is a logged in backend user.  If there is, do the redirect to get a new token. If not, ditch the current token and reload (Although I'd be unhappy with a real reload, maybe Apollo Client could just try to connect a second time )

## Steps to Reproduce
Please provide starting context, i.e. logged in as a user, configure a particular payment method.
1. Log in to the storefront
2. Wait until the token expires
2.5 Im not sure when actually the backend logout happens
3. Visit the storefront again, and see that it will redirect to the backend login form.

## Possible Solution
I'm unsure if that can be fixed in the backend on the route that also provides the silent reauth, or has to be done in the storefront.

We currently don't separate actual logins from silent reauths in the `/login`endpoint on the Backend:
https://github.com/reactioncommerce/reaction/blob/746f5acf8eb2a9e877d7ede93f46a595b8013348/imports/plugins/core/hydra-oauth/server/oauthEndpoints.js#L18

We could instead provide both a `/login` and `/refresh` (or `/reauth` as it isn't really a refresh in oauth terminology) endpoint. While the behaviour for both would be the same when `getLoginRequestRes.skip` is set, for the other case `/login` could still redirect to the login form as usual, while `/refresh` would redirect back to the storefront instead. I The storefront would have dropped the session cookie by then so it won't try the reauth again.

I think it would be even better if the unsuccessful redirect could be completely avoided, as it would happen quite often for logged in users. But the storefront would need to know beforehand if there is an authenticated backend user, and I'm not sure if that could work, as I think the Meteor Client bundle will always need to load once and grab the token from `localStorage` to which no other page has access. A hidden iframe solution could maybe work, that would only trigger a redirect if silent re-authentication is possible.

Im still need to figure out why exactly this issue occurs where the backend doesn't have valid authentication anymore while the frontend still thinks it is logged in.

### Versions
2.1.0

**Comment by aldeed** [comment]
@janus-reith Moved the issue to this repo, thanks.

This is great investigation work. It seems to mostly align with what I was thinking.
- Your implementation sounds correct to me, so a PR would be great.
- Getting a new refresh token with every refresh is how it always works in oauth as far as I remember. It should be updated and used along with the auth token. That way you will remain logged in (by default) for 30 days from last refresh, rather than 30 days from first login.
- If you need to test more quickly, you can set `TTL_ACCESS_TOKEN` and `TTL_REFRESH_TOKEN` env on Hydra service to smaller values. I believe these are both set in number of seconds.
- Without a page refresh, I think it would get a new access token on the client, but the server would have the old one, which could lead to weirdness on the next user-initiated refresh. Maybe not, but it's something to check.
- Make sure whatever you do works correctly when `ENABLE_SPA_ROUTING` env is set to `true` and also when set to `false`. With SPA routing, most route changes are intercepted on the client. Without it, all route changes do a full page refresh.

**Comment by janus-reith** [comment]
@aldeed Thanks, will keep on testing and prepare a PR.
Had this in use in a dev environment for the past week and didn‘t encounter any errors yet.

Regarding the refresh token: I think it should only issue a new one when necessary. The (semi-)critical part I see here also is not the issuing of a new refresh token, but the invalidation of the previous one. If I’m logged in with client a and client b and both share the same login, as soon as client a would refresh, the token of client b would be considered invalid and client b would be logged out on the next visit after being redirected to my refresh endpoint.
Client B might have an outdated Access Token, but the refresh token should still be valid to allow for a token refresh.

However, I don’t consider this a blocker and would still favor this over the current implementation, also I think that issue is not related to the changes supposed here, as the token endpoint I call in my implementation seems to be the correct one, for reference: https://www.oauth.com/oauth2-servers/access-tokens/refreshing-access-tokens/
Still something to keep in mind and research.

**Comment by aldeed** [comment]
@janus-reith Is this still on your "make a PR when I get a chance" list?

**Comment by janus-reith** [comment]
@aldeed Hi, sorry, yes it is, altough I focused on the storefront reimplementation and fixed it there in one go.
I`m done with most of the stuff there and could quickly do a PR for the token refresh here if this still is relevant,

**Comment by aldeed** [comment]
It would be nice to fix here if it’s not a lot of work.

**Comment by janus-reith** [comment]
Will prepare something

**Comment by janus-reith** [comment]
Just to give an update, I basically finished this,
but faced a weird error here that I couldn't resolve yet properly:
https://github.com/reactioncommerce/example-storefront/blob/trunk/src/lib/apollo/initApollo.js#L68-L72

I had to remove `options.req.logout()`, as now the /refresh route will determine wether a refresh was successful, or the user should be logged out to prevent an infinite loop.
The weird thing: If that predecending `logout` call on `req` is not present, and I call `options.res.writeHead`, the app will crash, stating "`Cannot set headers after they are sent to the client`". This is however expected to only execute if `process.browser` is false.

This is not related to my implementation, but also is present on the default one when you remove the `options.req.logout()` part. The expectation would then be, that the app may take an infinite loop of trying to log in, but instead it will crash right away with that error.
I think this is somehow related to the fact that multiple Apollo Queries are executing in parallel, and this leads me to believe that this redirecting in there should never have worked in the first place. I'm really curious why that call on `req` migitates that error, I don't think it is because the actual logout.
I tried several things to fix this, but without success, the closest I could get was the wrap this in a try-catch block so only the first executes properly, but I don't really consider this a solution.

In my own usecase for the new storefront I dont handle any authenticated queries on SSR Render anymore and therefore didn't have that happening for me before, as only the `Router.pushRoute` clientside routing is left and I also believe that to be the way to go.

What I'm trying now is to get rid of the redirect alltogether and make this a completely silent refresh implementation, by doing a fetch call to the refresh endpoint inside the `onError` Link, that endpoint would now return the new acessToken as JSON instead of a redirect, rest is the same, new tokens including refresh token are storen encrypted in the cookie.
But that turned out to be really complex: The default example in the apollo docs doesn't seem to mimic a real usecase, as this only works as a sync call, which wouldn't work for fetching tokens.

What I need to do is to stall the query in there, try a token refresh, then execute it, either with a new token or unauthenicated without one.
I also found a solution for that - But the next issue is that by default, the storefront will have around 3-4 Queries executing in a short time frame, and I don't want each of them to trigger another token refresh plus the added delay of another fetch, in case the previous call doing that didn't finish yet.

So I have to batch all pending Queries, wait for the token refresh, then execute them, and I still have some issues at this point.
I still think it is worth to figure this out, as the new storefront would also benefit from the silent refresh instead of a clientside redirect.

I'm really baffled BTW that I couldn't find any official or complete example on how to handle the refresh properly in Apollo Client.

**Comment by aldeed** [comment]
@janus-reith Seems like you've put a lot of work into this, so thank you for that. Some of your issues bring to mind the same roadblocks we hit in the initial implementation.

I haven't looked at it closely, but there might be some techniques in this package that could help: https://www.npmjs.com/package/apollo-link-token-refresh

The logout call does nothing except clear the auth cookie. I would need to debug a lot more to fully understand how the multiple requests are clashing with each other, but it makes sense that would happen. Attempting to track the expiration time and reauth _before_ the token expires (i.e., not in `onError`) might be an option, but it might miss some situations.

You might try posting an issue with some of this info in the Apollo Client repo. I'm sure they'd love to have good docs on how to do this with a standard OAuth flow, which they don't currently have. Maybe some minimal changes to the core packages would simplify what you're trying to do. It also seems like a situation where creating a minimal reproduction app would be helpful. An app that just has a single page that looks different depending on user and tries to fulfill all the requirements of the storefront.

**Comment by janus-reith** [comment]
@aldeed Thanks, I also noticed that package, and I think it basically does the same thing, which is to queue upcoming queries until the fetch is resolved.
I will compare implementations, because that's where I got stuck and didn't have time left last time. Also I did't want to just plug that in, as I didn't want to introduce another unknown here and instead want to be able to actually keep track of what happens at what point in the chain.
